middleware/csrf/memory.go

```diff
@@ -0,0 +1,78 @@
+package csrf
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/gofiber/fiber/v2/utils"
+)
+
+type memoryStorage struct {
+	sync.RWMutex
+	tokens map[string]int64
+}
+
+// NewMemoryStorage - Creates new in-memory storage for CSRF tokens
+func NewMemoryStorage() fiber.Storage {
+	storage := &memoryStorage{
+		tokens: make(map[string]int64),
+	}
+	go storage.gc()
+	return storage
+}
+
+func (m *memoryStorage) Get(id string) ([]byte, error) {
+	m.RLock()
+	t, ok := m.tokens[id]
+	m.RUnlock()
+	// Check if token exist or expired
+	if !ok || time.Now().Unix() >= t {
+		return nil, errors.New("csrf: invalid key")
+	}
+
+	return utils.GetBytes(id), nil
+}
+
+// Set session value. `exp` will be zero for no expiration.
+func (m *memoryStorage) Set(id string, value []byte, exp time.Duration) error {
+	m.Lock()
+	m.tokens[id] = time.Now().Unix() + int64(exp.Seconds())
+	m.Unlock()
+
+	return nil
+}
+
+// Delete session value
+func (m *memoryStorage) Delete(id string) error {
+	m.Lock()
+	delete(m.tokens, id)
+	m.Unlock()
+	return nil
+}
+
+// Clear clears the storage
+func (m *memoryStorage) Clear() error {
+	m.Lock()
+	for k := range m.tokens {
+		delete(m.tokens, k)
+	}
+	m.Unlock()
+	return nil
+}
+
+func (m *memoryStorage) gc() {
+	for {
+		// GC the tokens every 10 seconds to avoid
+		time.Sleep(10 * time.Second)
+		now := time.Now().Unix()
+		for t := range m.tokens {
+			if now >= m.tokens[t] {
+				m.Lock()
```

Comment:
I'm wondering if we don't put this outside the for-range, will we have a race condition? @Fenny

Reply:
i think not, reading from map with multiple goroutines should be safe, write lock will not allow reads from map, so as long as writes are safe, this will not be race condition

Reply:
It's really bad for performance to acquire a lock every 10 seconds, maybe a Read lock, but i dont think this is important, memory provider should not be used in production anyway, scaling is impossible